File: android/src/main/java/cn/genee/rn/serialport/RNSerialPort.java
package cn.genee.rn.serialport;

import android.content.res.AssetFileDescriptor;
import android.content.res.AssetManager;
import android.database.Cursor;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Environment;
import android.os.StatFs;
import android.provider.MediaStore;
import android.util.Base64;
import android.util.SparseArray;
import android.media.MediaScannerConnection;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.module.annotations.ReactModule;
import com.facebook.react.modules.core.RCTNativeAppEventEmitter;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.RandomAccessFile;
import java.net.URL;
import java.security.MessageDigest;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

@ReactModule(name = RNSerialPort.MODULE_NAME)
public class RNSerialPort extends ReactContextBaseJavaModule {

  static final String MODULE_NAME = "RNSerialPort";

  private ReactApplicationContext reactContext;

  public RNSerialPort(ReactApplicationContext reactContext) {
    super(reactContext);
    this.reactContext = reactContext;
  }

  private void sendEvent(String eventName,
  @Nullable WritableMap params) {
getReactApplicationContext()
.getJSModule(RCTNativeAppEventEmitter.class)
.emit(eventName, params);
}

private void sendEventWithJson(String eventName,
          JSONObject json) {
try {
WritableMap map = JsonConvert.jsonToReact(json);
sendEvent(eventName, map);
} catch (JSONException ex) {
Log.d(LOG_TAG, "fireNdefEvent fail: " + ex);
}
}


  @Override
  public String getName() {
    return MODULE_NAME;
  }

  @ReactMethod
  public void getFilePort(String path, int baudRate, Promise promise) {
    return 1;
  }

  @ReactMethod
  public void getUSBPort(int vendorId, int productId, int baudRate, Promise promise) {
    return 2;
  }

}

File: android/src/main/java/cn/genee/rn/serialport/RNSerialPort.kt
package cn.genee.rn.serialport

import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule

import java.net.URI
import java.util.concurrent.Executors

class RNSerialPort(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {

    private fun sendEvent(
            deviceId: String,
            params: WritableMap
    ) {
        if (reactApplicationContext.hasActiveCatalystInstance()) {
            reactApplicationContext
                    .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                    .emit("SerialPort.event@$deviceId", params)
        }
    }

    override fun getName() = "RNSerialPort"

    override fun onCatalystInstanceDestroy() {
        RNLog.d("onCatalystInstanceDestroy: close all opening ports")
        ports.forEach { (_, port) ->
            port.close()
        }
        ports.clear()
        super.onCatalystInstanceDestroy()
    }

    @ReactMethod
    fun openPort(deviceId: String, baudRate: Int) {
        if (!ports.containsKey(deviceId)) {
            RNLog.d("openPort deviceId=$deviceId baudRate=$baudRate")
            val uri = URI(deviceId)
            when (uri.scheme) {
                "file" -> {
                    RNLog.d("openFilePort path=${uri.path}")
                    ports[deviceId] = object : FileSerialPort(reactApplicationContext, uri.path, baudRate) {
                        override fun onData(data: ByteArray) {
                            val params = WritableNativeMap()
                            params.putString("event", "data")

                            val dataArray = WritableNativeArray()
                            data.map { dataArray.pushInt(it.toInt()) }
                            val eventParams = WritableNativeArray()
                            eventParams.pushArray(dataArray)
                            params.putArray("params", eventParams)

                            sendEvent(deviceId, params)
                        }

                        override fun onClose() {
                            val params = WritableNativeMap()
                            params.putString("event", "close")
                            sendEvent(deviceId, params)
                            ports.remove(deviceId)
                        }
                    }.apply {
                        executor.execute(this)
                    }
                }
                "usb" -> {
                    RNLog.d("openUSBPort id=${uri.path}")
                    ports[deviceId] = object : USBSerialPort(reactApplicationContext, uri.path.toInt(), baudRate) {
                        override fun onData(data: ByteArray) {
                            val params = WritableNativeMap()
                            params.putString("event", "data")

                            val dataArray = WritableNativeArray()
                            data.map { dataArray.pushInt(it.toInt()) }
                            val eventParams = WritableNativeArray()
                            eventParams.pushArray(dataArray)
                            params.putArray("params", eventParams)

                            sendEvent(deviceId, params)
                        }

                        override fun onClose() {
                            val params = WritableNativeMap()
                            params.putString("event", "close")
                            sendEvent(deviceId, params)
                            ports.remove(deviceId)
                        }
                    }.apply {
                        executor.execute(this)
                    }
                }
            }
        }
    }

    @ReactMethod
    fun closePort(deviceId: String) {
        RNLog.d("closePort deviceId=$deviceId")
        if (ports.containsKey(deviceId)) {
            ports[deviceId]?.close()
            ports.remove(deviceId)
        }
    }

    @ReactMethod
    fun writePort(deviceId: String, baudRate: Int, bytesArray: ReadableArray, promise: Promise) {
        val bytes = ByteArray(bytesArray.size()) {
            bytesArray.getInt(it).toByte()
        }

        openPort(deviceId, baudRate)

        val port = ports[deviceId]
        if (port !== null) {
            executor.execute {
                RNLog.d("writePort deviceId=$deviceId size=${bytes.size}")
                port.write(bytes)
                promise.resolve(bytes.size)
            }
        } else {
            promise.resolve(-1)
        }
    }

    companion object {
        private val ports = HashMap<String, SerialPort>()
        private val executor = Executors.newCachedThreadPool()
    }
}
